AddOn - AWS Secrets Manager and Config Provider for Secret Store CSI Driver

anshrma opened the pull request:
### What does this PR do?

This PR adds two Kubernetes add-on , (1) AWS Secrets Manager and Config Provider for Secret Store CSI Driver (2) Secrets Store CSI Driver 


### Motivation

AWS partner needing this add on to complete their development work.
Closes #435 

### More

- [ Y] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ Y ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ NA ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ Y] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR


**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

<!-- Anything else we should know when reviewing? -->

vara-bonthu commented on modules/kubernetes-addons/csi-secrets-store-provider-aws/locals.tf:
```diff
@@ -0,0 +1,32 @@
+    timeout     = "1200"
+  }
+
+  default_helm_values = []
```
you can remove this if you are setting as empty list

vara-bonthu commented on modules/kubernetes-addons/csi-secrets-store-provider-aws/locals.tf:
```diff
@@ -0,0 +1,32 @@
+    version     = "0.0.2"
+    namespace   = local.name
+    description = "A Helm chart to install the Secrets Store CSI Driver and the AWS Key Management Service Provider inside a Kubernetes cluster."
+    values      = local.default_helm_values
```
change this to `values = []`

vara-bonthu commented on modules/kubernetes-addons/secrets-store-csi-driver/locals.tf:
```diff
@@ -0,0 +1,36 @@
+    timeout     = "1200"
+  }
+
+  default_helm_values = []
```
remove it same as above comment

vara-bonthu reviewed (approved):
Great work @anshrma ! LGTM 👍🏼  

Few minor changes requested. We just need to run some E2E tests before merging the PR

anshrma commented:
> Great work @anshrma ! LGTM 👍🏼
> 
> Few minor changes requested. We just need to run some E2E tests before merging the PR

Thanks @vara-bonthu  - Addressed your feedbacks.

naris-silpakit reviewed (approved):
LGTM!

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/README.md:
```diff
@@ -0,0 +1,131 @@
+# EKS Cluster Deployment with new VPC
```
Instead of creating a whole new example just for these two addons, instead can we just add their implementation into the https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/complete-kubernetes-addons

bryantbiggs replied:
@askulkarni2 / @kcoleman731 what are your thoughts on this as we continue to expand the number of addons?

vara-bonthu replied:
I agree with @bryantbiggs . We should start moving the individual Add-on examples to `[complete-kubernetes-addons](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/complete-kubernetes-addons)` . This will help us to manage limited exmaples

vara-bonthu replied:
We can add some more details to `docs` `.md` files as well.  New users should be able to goto docs to findout the usage of this Add-on.

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/data.tf:
```diff
@@ -0,0 +1,11 @@
+data "aws_iam_policy_document" "secrets_management_policy" {
```
when you move this over to the `complete-kubernetes-addons` example, no need to put this in a separate file. Instead, locate it next to where it is used. Here that would be either before or after https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/471/files#diff-6355537d0664dc024a0086c9581f3a707332dd67c563f98cbe5ee18d4fd0e44bR164

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/main.tf:
```diff
@@ -0,0 +1,268 @@
+# Kubernetes CRD to create the "SecretProviderClass" to represent the Secrets Manager Secrets
+#---------------------------------------------------------------
+
+resource "kubernetes_manifest" "csi_secrets_store_crd" {
```
we should use [`kubectl_manifest`](https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/resources/kubectl_manifest) in the examples to avoid hitting this issue when deploying cluster and addons https://github.com/hashicorp/terraform-provider-kubernetes/issues/1453

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/main.tf:
```diff
@@ -0,0 +1,268 @@
+# Sample Kubernetes Pod to mount the Secrets as CSI Volume
+#---------------------------------------------------------------
+
+resource "kubernetes_manifest" "sample_nginx" {
```
same as above, lets used `kubectl_manifest` instead for now in examples

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/secretconfig.yaml:
```diff
@@ -0,0 +1,11 @@
+# This file represents `spec.parameters.objects` of [SecretProviderClass]
+# (https://docs.aws.amazon.com/secretsmanager/latest/userguide/integrating_csi_driver.html#integrating_csi_driver_SecretProviderClass)
+- objectName: "ARN OF THE SECRET OBJECT, which has username and password as objects as example"
```
to test this we should create a secret and show that  its actually mounted and retrievable from within a pod. we can create the secrets manager secret within the terraform example and we'll need to use the [random provider](https://registry.terraform.io/providers/hashicorp/random/latest/docs) to generate a random name for the secret since secrets are unique per account/region and they stay around for a minimum of 7 days pending deletion

anshrma replied:
Updated the examples to create secret on the fly. No need to use the random provider. With Terraform destroy, the secret also gets deleted.

bryantbiggs commented on modules/kubernetes-addons/csi-secrets-store-provider-aws/main.tf:
```diff
@@ -0,0 +1,15 @@
+resource "kubernetes_namespace" "csi_secrets_store_provider_aws" {
```
I believe we can remove this and instead use whats provided in the `helm_addon` sub-module https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/deec7d5caea47c06dea48fa616ad2c56e52c3cce/modules/kubernetes-addons/helm-addon/main.tf#L64

askulkarni2 replied:
Or alternatively set `create_kubernetes_namespace` and `create_kubernetes_service_account` to `false` in the `module helm_addon` call.

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/main.tf:
```diff
@@ -0,0 +1,268 @@
+# Creating IAM Role for Service Account
+#---------------------------------------------------------------
+
+module "iam_role_service_account" {
```
I believe this can all be removed and the addon should be creating this https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/deec7d5caea47c06dea48fa616ad2c56e52c3cce/modules/kubernetes-addons/helm-addon/main.tf#L61

Is that correct @vara-bonthu ?

askulkarni2 replied:
So, the problem here is that the addon itself has no clue what secrets are being retrieved. The way it works, secrets are retrieved by the AWS provider when a pod needing those secrets is started and removed afterwords. The service-account associated with the respective pods needs to have the needed IRSA config. These policies cannot be added to the addon. In CDK, we do this when the `Team` or `AddOn` secrets are defined. The same thing needs to happen here. That is to say, in order to consume secrets, there are two paths:

1. You need to define your secrets via `Teams` configuration (similar to how we do labels and quotas today and we will have to update the Teams module for this). We need to add the required policies to the IRSA config we define for the Team.
2. For add-ons, this will be a parameter for the add-on in which case we will add the required policies to the addons IRSA config.

CC: @kcaws @shapirov103 I am assuming we want to keep the design for secrets somewhat similar between the two projects.

anshrma replied:
@askulkarni2 / @kcaws - Though i agree on abstracting out IRSA specifics out in a teams configuration, but that work is pretty involved and hence i propose that we proceed with reviewing this PR and merging it and consider the teams configuration with a separate PR.  The implementation details of the module wont change and changes developed in this PR form the basis of next one (+ also provides customers an example on implementation without Teams configuration similar to CDK). I have addressed all the comments otherwise.

bryantbiggs commented on modules/kubernetes-addons/secrets-store-csi-driver/locals.tf:
```diff
@@ -0,0 +1,34 @@
+  set_values = [
+    {
+      name  = "enableSecretRotation"
+      value = "true"
```
what is the motivation for hardcoding this as true for users?

askulkarni2 replied:
+1. Leave this out, let users pick via `values.yaml`

bryantbiggs commented on modules/kubernetes-addons/secrets-store-csi-driver/locals.tf:
```diff
@@ -0,0 +1,34 @@
+    },
+    {
+      name  = "syncSecret.enabled"
+      value = "true"
```
Likewise, what is the motivation for hardcoding this as true for users?

bryantbiggs commented on modules/kubernetes-addons/secrets-store-csi-driver/main.tf:
```diff
@@ -0,0 +1,15 @@
+resource "kubernetes_namespace" "secrets_store_csi_driver" {
```
see other addon, should be handled by `helm_addon` I believe

askulkarni2 replied:
Same as above.

bryantbiggs commented on docs/add-ons/csi-secrets-store-provider-aws.md:
````diff
@@ -0,0 +1,12 @@
+csi-secrets-store-provider-aws can be deployed by enabling the add-ons via the following.
+
+```hcl
+enable_secrets_store_csi_driver = true
````
question - should these be two separate addons, or just one addon? If I can't run just the `csi-secrets-store-provider-aws` addon by itself then there is a hard dependency between the two addons

Thoughts @askulkarni2 / @kcoleman731 / @vara-bonthu ?

bryantbiggs commented on examples/secrets-management-with-csi-secrets-driver-aws/README.md:
```diff
@@ -0,0 +1,131 @@
+
+#### Step7: Exec in the Pod to view the secrets getting mounted successfully
+
+    $ kubectl exec -it -n nginx pod/nginx-secrets-pod-sample bash
```
how is this pod deployed? and conversely, what secrets are we expecting to see if the [secretconfig](https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/471/files#diff-948ff2266135aa06316f7d25374d5c4ef3eda95c9fab5827fb2f305caa099a4bR3) does not contain any valid secret ARNs?

anshrma replied:
Pod is deployed in the example itself as a sample nginx container. I removed the need of secretconfig now in the sample as i am creating the secret in the example itself.

bryantbiggs reviewed (changes requested):
Thanks for the PR! Please update your branch with the latest on `main` as there have been a number of changes to the `examples/` format

askulkarni2 reviewed (commented):
First of all thanks so much for this PR! The PR is looking good. There is one main outstanding design item. We probably want to keep the secrets provisioning design similar to how we do in CDK (if possible). Refer to https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/docs/extensibility.md#passing-secrets-to-add-ons.

vara-bonthu commented on modules/kubernetes-addons/secrets-store-csi-driver/locals.tf:
```diff
@@ -0,0 +1,30 @@
+    var.helm_config
+  )
+
+  irsa_config = {
+    create_kubernetes_namespace       = true
+    kubernetes_namespace              = local.name
+    create_kubernetes_service_account = false
+    kubernetes_service_account        = local.name
+  }
```
Don't use irsa module if we are not creating an IAM role with service account. 
Simply create a Namespace object for this

vara-bonthu commented on modules/kubernetes-addons/csi-secrets-store-provider-aws/locals.tf:
```diff
@@ -0,0 +1,30 @@
+    var.helm_config
+  )
+
+  irsa_config = {
+    create_kubernetes_namespace       = true
+    kubernetes_namespace              = local.name
+    create_kubernetes_service_account = false
+    kubernetes_service_account        = local.name
+  }
```
Don't use irsa module if we are not creating an IAM role with service account. 
Simply create a Namespace object for this

anshrma replied:
@vara-bonthu - I was creating a namespace object earlier, but i was asked to use this instead per https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/471#discussion_r863312265 . Tagging @askulkarni2 for visibility. Pls lmk a final direction that you both align on.

vara-bonthu replied:
@anshrma Please go-head and change this to simple namespace resource. Use IRSA module only when you need IAM role for service accounts

askulkarni2 reviewed (commented):
@anshrma looking good, just one final change for using namespace resource instead of irsa module as suggested by Vara.

anshrma commented:
> @anshrma looking good, just one final change for using namespace resource instead of irsa module as suggested by Vara. Also please address the CI failures. Running `pre-commit run -a` should fix most of them.

Thanks @askulkarni2 . Done with the changes requested and addressed the CI failures. Please let me know, if any other changes required.

BlZvi commented:
Hey @anshrma, @askulkarni2 and @bryantbiggs, Whaat is missing in this PR to go forward?

askulkarni2 reviewed (approved):
Updated, tested and approved!
